ironfish/src/wallet/wallet.ts

```diff
@@ -1292,18 +1293,34 @@ export class Wallet {
       throw new Error(`Account already exists with the name ${toImport.name}`)
     }
 
-    if (this.listAccounts().find((a) => toImport.spendingKey === a.spendingKey)) {
+    if (
+      toImport.spendingKey &&
+      this.listAccounts().find((a) => toImport.spendingKey === a.spendingKey)
+    ) {
       throw new Error(`Account already exists with provided spending key`)
     }
 
-    const key = generateKeyFromPrivateKey(toImport.spendingKey)
-
-    const accountValue: AccountValue = {
-      ...toImport,
-      id: uuid(),
-      incomingViewKey: key.incoming_view_key,
-      outgoingViewKey: key.outgoing_view_key,
-      publicAddress: key.public_address,
+    let accountValue: AccountValue = {} as AccountValue
+    if (toImport.spendingKey) {
+      // if spending key is provided, derive everything from that, even if view keys were also provided.
+      const key = generateKeyFromPrivateKey(toImport.spendingKey)
+      accountValue = {
+        ...toImport,
+        id: uuid(),
+        incomingViewKey: key.incoming_view_key,
+        outgoingViewKey: key.outgoing_view_key,
+        publicAddress: key.public_address,
+      }
+    } else if (toImport.incomingViewKey && toImport.outgoingViewKey) {
+      // if spending key is not provided, use the provided view keys
+      const publicAddress = generatePublicAddressFromIncomingViewKey(toImport.incomingViewKey)
+      accountValue = {
+        ...toImport,
+        id: uuid(),
+        incomingViewKey: toImport.incomingViewKey,
+        outgoingViewKey: toImport.outgoingViewKey,
```

Comment:
these two lines are unnecessary

Reply:
thought so too, but an error occurs without them:

```
Type '{ id: string; publicAddress: string; name: string; spendingKey?: string | undefined; incomingViewKey?: string | undefined; outgoingViewKey?: string | undefined; }' is not assignable to type 'AccountValue'.
  Types of property 'incomingViewKey' are incompatible.
    Type 'string | undefined' is not assignable to type 'string'.
      Type 'undefined' is not assignable to type 'string'.ts(2322)
```

Reply:
Actually, with the type change suggested elsewhere, no error B)

Reply:
Yea you can't do assertions of object properties and have the type system pick that up unfortunately, you have to unpack and do assertions on them. Either way this should be a moot point if we do the OR type rather than optional values